types/src/shared/cache.ts

```diff
@@ -31,15 +34,25 @@ export function cacheWithRedis<T extends (...args: any[]) => Promise<any>>(
         redisUri,
       });
       const key = `cacheWithRedis-${fn.name}-${resolver(...args)}`;
-      const cacheVal = await redisCli.get(key);
+      let cacheVal = await redisCli.get(key);
```

Comment:
I might not have got this properly, but why not leverage Redis all the way?

Something like this. Yet not perfect as it relies on a setTimeout to poll every X ms see if the lock can be acquired.
```
async function getCachedValue(key: string): Promise<string> {
  const lockKey = `${key}_lock`;

  // First attempt to retrieve the cached value without acquiring the lock.
  const ley = ...
  const cachedValue = await redisClient.get(key);
  if (cachedValue) {
    return cachedValue;
  }

  // If not cached, acquire the lock before computing.
  while (true) {
    const lockAcquired = await redisClient.set(lockKey, 'locked', { NX: true, EX: 10 });
    if (lockAcquired) {
      try {
        // Double-check if another process might have cached the value while acquiring the lock.
        const cachedValueAfterLock = await redisClient.get(key);
        if (cachedValueAfterLock) {
          return cachedValueAfterLock;
        }

        const computedValue = await fn(...args);
        await redisCli.set(key, JSON.stringify(result), {
          PX: ttlMs,
        });
        
        return computedValue;
      } finally {
        await redisClient.del(lockKey);
      }
    } else {
      // If the lock is held by another process, wait before retrying.
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
  }
}
```

Reply:
Seeing it the other way round: why use redis?

as you said we need polling in a `while(true)` which i'm not a fan of

Reply:
Answered on first comment.